rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/LocalInfrastructure.java

```diff
@@ -267,27 +290,32 @@ protected void configure(Object... args) {
      */
     @Override
     protected void notifyDeployingNodeLost(String pnURL) {
-        // notifies the watching loop to exit... it will clean
-        // isDeployingNodeLost & isNodeAcquired collections up.
-        this.isDeployingNodeLost.put(pnURL, true);
+        this.lostNodes.incrementAndGet();
     }
 
     @Override
     protected void notifyAcquiredNode(Node arg0) throws RMException {
-        this.isNodeAcquired.put(arg0.getNodeInformation().getName(), true);
+        this.acquiredNodes.incrementAndGet();
     }
 
     @Override
     public void removeNode(Node node) throws RMException {
-        String nodeName = node.getNodeInformation().getName();
-        logger.debug("Removing node " + node.getNodeInformation().getURL() + " from infrastructure");
-        this.atomicMaxNodes.incrementAndGet();
-        ProcessExecutor proc = this.nodeNameToProcess.remove(nodeName);
-        if (proc != null) {
-            proc.killProcess();
-            logger.info("Process associated to node " + nodeName + " destroyed");
-        } else {
-            logger.warn("No process associated to node " + nodeName);
+        logger.debug("Removing node " + node.getNodeInformation().getURL() + " from " + this.getClass().getSimpleName());
+
+        if (!this.nodeSource.getDownNodes().contains(node)) {
+            // the node was manually removed
+            handledNodes.decrementAndGet();
+        }
+
+        int remainingNodesCount = this.acquiredNodes.decrementAndGet();
+        // If there is no remaining node, kill the JVM process
+        if (remainingNodesCount == 0) {
+            if (processExecutor != null) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Merge this if statement with the enclosing one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1066)